benchexec/container.py

```diff
@@ -746,6 +779,77 @@ def escape(s):
     )
 
 
+def check_use_fuse_overlayfs(mount_base, dir_modes):
+    mount_points = [
+        (full_mountpoint, fstype)
+        for _unused_source, full_mountpoint, fstype, _options in get_mount_points()
+        if util.path_is_below(full_mountpoint, mount_base)
+    ]
+
+    for full_mountpoint, fstype in mount_points:
+        mountpoint = full_mountpoint[len(mount_base) :] or b"/"
+        mode = determine_directory_mode(dir_modes, mountpoint, fstype)
+
+        if not mode or not os.path.exists(mountpoint):
+            continue
+
+        if mode == DIR_OVERLAY:
+            # Check if there are any sub-mounts within the current overlay mount point
+            for sub_mountpoint, _unused_fstype in mount_points:
+                if (
+                    util.path_is_below(sub_mountpoint, mountpoint)
+                    and sub_mountpoint != mountpoint
+                ):
+                    return True
+
+    return False
+
+
+def fuse_overlay_mount_root(fuse, temp_base, work_base):
+    """
+    Mount "/" once with fuse-overlayfs once, use it for all overlay mounts.
+    """
+    temp_fuse = temp_base + b"/fuse"
+    work_fuse = work_base + b"/0"
+    os.makedirs(temp_fuse, exist_ok=True)
+    os.makedirs(work_fuse, exist_ok=True)
+    make_fuse_overlay_mount(fuse, temp_fuse, b"/", temp_base, work_fuse)
+
+
+def make_fuse_overlay_mount(exe, mount, lower, upper, work):
+    logging.debug(
+        "Creating overlay mount with fuse-overlayfs: target=%s, lower=%s, upper=%s, work=%s",
+        mount,
+        lower,
+        upper,
+        work,
+    )
+
+    cmd = (
+        exe,
+        b"-o",
+        b"lowerdir="
+        + escape(lower)
+        + b",upperdir="
+        + escape(upper)
+        + b",workdir="
+        + escape(work),
+        escape(mount),
+    )
+
+    try:
+        cap_permitted_to_ambient()
+        result = subprocess.run(
+            args=cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+        )
+        if result.stdout:
+            logging.debug("fuse-overlayfs: %s", result.stdout.decode())
+    except subprocess.CalledProcessError as e:
+        sys.exit(f"Error executing command: {e}\n{e.stdout.decode()}")
+    finally:
+        drop_ambient_cap()
```

Comment:
I just realized: With `@contextmanager` from `contextlib` we could write a single function `permitted_cap_as_ambient()` that could be used as a context manager:
```python
with permitted_cap_as_ambient():
     result = subprocess.run(...)
```
I think this is easy to do, just have a look at the example in the documentation of `@contextmanager`. And it makes it impossible to forget the dropping of capabilities.

Would you like to add this?

Reply:
Cool, I've done that. I'll submit a new patch.